Healthcare-Hospital-Management-System/Healthcare-Hospital-Management-System/Infrastructure/DrugClient.cs

```diff
@@ -0,0 +1,94 @@
+﻿using System.Text.Json;
+
+namespace HealthcareHospitalManagementSystem.Services
+{
+    public class DrugClient
+    {
+        private readonly HttpClient _httpClient;
+
+        public DrugClient(HttpClient httpClient)
+        {
+            _httpClient = httpClient;
+        }
+
+        public async Task<DrugReportClass> GetDrugReportAsClassAsync(string searchTerm, CancellationToken cancellationToken = default)
+        {
+            string apiKey = Environment.GetEnvironmentVariable("OPENFDA_API_KEY");
+            if (string.IsNullOrEmpty(apiKey))
+            {
+                throw new InvalidOperationException("API key not found. Please set the 'OPENFDA_API_KEY' environment variable.");
+            }
+            var jsonData = await FetchApiData(searchTerm, apiKey, cancellationToken);
+            return ParseDataAsClass(jsonData);
```

Comment:
This can be done in a more efficient way, you do not need ParseDataAsClass() if you are using ReadFromJsonAsync<T>()
https://learn.microsoft.com/en-us/dotnet/api/system.net.http.json.httpcontentjsonextensions.readfromjsonasync?view=net-8.0
```
public async Task<DrugReportClass?> GetDrugReportAsClassAsync(string searchTerm, CancellationToken cancellationToken = default)
{
    string apiKey = Environment.GetEnvironmentVariable("OPENFDA_API_KEY");
    if (string.IsNullOrEmpty(apiKey))
    {
        throw new InvalidOperationException("API key not found. Please set the 'OPENFDA_API_KEY' environment variable.");
    }
    return await FetchApiData(searchTerm, apiKey, cancellationToken);
}

private async Task<DrugReportClass?> FetchApiData(string searchTerm, string apiKey, CancellationToken cancellationToken)
{
    var requestUrl = $"https://api.fda.gov/drug/event.json?search={searchTerm}&api_key={apiKey}";
    var response = await _httpClient.GetAsync(requestUrl, cancellationToken);
    response.EnsureSuccessStatusCode();
    return await response.Content.ReadFromJsonAsync<DrugReportClass>(cancellationToken);
}
```

Reply:
If you still would like to use ReadAsStringAsync(), you still do not need ParseDataAs...() and Custom implementation of Json deserializer. You can use methods from the box like JsonSerializer.Deserialize<DrugReportClass>(jsonData);
https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/how-to

```
public async Task<DrugReportClass?> GetDrugReportAsClassAsync(string searchTerm, CancellationToken cancellationToken = default)
{
    string apiKey = Environment.GetEnvironmentVariable("OPENFDA_API_KEY");
    if (string.IsNullOrEmpty(apiKey))
    {
        throw new InvalidOperationException("API key not found. Please set the 'OPENFDA_API_KEY' environment variable.");
    }
    var jsonData = await FetchApiData(searchTerm, apiKey, cancellationToken);
    return JsonSerializer.Deserialize<DrugReportClass>(jsonData);
}
```